Source/ARTRest.m

```diff
@@ -742,6 +742,19 @@ - (ARTLocalDevice *)device_nosync {
     return _sharedDevice;
 }
 
+/*
+ This function is needed to return functioning local device, since after `reset` call
```

Comment:
`-[ARTRest device]` is still part of the public interface of the library, so I think that method needs to always return a "functioning"  local device without needing to first call `loadDevice` (which isn't even a part of the public interface). Could we not move the logic of "if the device has been reset, then create a new one" into `-[ARTRest device_noSync]`?

Reply:
I tried that, but it immediately re-creates the device and thus it's never can be in a `isReset` state which was the whole point of these changes and underlying issue.

Reply:
I think that the issue being described by #1177 actually consists of two separate things:

1. When the device deregisters itself for push notifications, it keeps its existing device details (`id`, `clientId`, etc), hence violating [RSH3g2a](https://docs.ably.io/client-lib-development-guide/features/#RSH3g2a).
2. Once we've fixed point 1, then we should make sure we respect [RSH8b](https://ably.com/documentation/client-lib-development-guide/features#RSH8b):

   > The `LocalDevice` `id` and `deviceSecret` attributes are generated, and persisted as part of the LocalDevice state, when required by step (RSH3a2b) in the Activation State Machine.

   That is, after the device has deregistered itself for push notifications, it should have "empty" (whatever that actually means) `id` and `deviceSecret` until the user tries to register the device for push notifications again. This is what @ben-xD showed happening on Android in his video.

So, I think that point 2 above is what you're referring to by "thus it's never can be in a `isReset` state", but I disagree that it "was the whole point of these changes" – we're also addressing point 1 above.

Furthermore, I think that trying to address point 2 as part of this PR brings additional complication, because we have _never_ respected RSH8b in our codebase. That is, at the moment, we generate the device `id` and `deviceSecret` the first time that `-[ARTRest device]` is called, instead of waiting until the user tries to register the device for push notifications. That's why we're now finding ourselves trying to answer questions like "should a reset device have an `id` of `@""`?" If we were already respecting RSH8b, then we'd already have answers to these questions.

I think that perhaps respecting RSH8b should be a separate task. And for me (possibly from lack of knowledge of the spec) it raises questions like those you've already identified – if there are API calls that require the device’s ID, which can be called before the push activation process is started, then where are we meant to get the ID from?

Reply:
Just to clarify to whoever will read this:

[RSH3g2a](https://docs.ably.io/client-lib-development-guide/features/#RSH3g2a) - reset device after deactivation (nullify all fields)
[RSH8b](https://docs.ably.io/client-lib-development-guide/features/#RSH8b) - re-initializing device's fields with the new `id` and `deviceSecret` on event `CalledActivate`.

Reply:
> I think that perhaps respecting RSH8b should be a separate task

Agreed

> if there are API calls that require the device’s ID, which can be called before the push activation process is started

I don't think so, it wouldn't make any sense. Or maybe I misunderstood what you've meant here. All functions that compose requests with the device `id` doesn't do any checks on it though:

`NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[[NSURL URLWithString:@"/push/deviceRegistrations"] URLByAppendingPathComponent:local.id]];`

Reply:
@lawrence-forooghian https://github.com/ably/ably-cocoa/issues/1256

Reply:
> Or maybe I misunderstood what you've meant here.

So, my comment was in response to [your code comment on `-[ARTRest localDevice]`](https://github.com/ably/ably-cocoa/blob/a88c282c0b88e58eb22adf7b272c49b6e649e264/Source/ARTRest.m#L746-L747):

>  This function is needed to return functioning local device, since after `reset` call
 you will end up with empty string as a device's `id` on the server.

I interpreted that comment as meaning the following: 

After calling `reset`, the `localDevice`'s `id` becomes an empty string. If we make an API request that uses the device's `id` whilst in this empty string state, then we'll submit an empty `id` to the server, hence we "will end up with empty string as a device's `id` on the server".

Did I misunderstand?

Reply:
> Did I misunderstand?

No, all is correct, but:

> so I think that method (`device`) needs to always return a "functioning" local device without needing to first call loadDevice

How is it possible to have an empty `id` then? It's either this or that ("reset" state or always "functioning"). That's why `loadDevice` exist - to load it on demand (when the function _needs_ it). Anyways lets further discuss it in a call.

Reply:
> if there are API calls that require the device’s ID, which can be called before the push activation process is started

Or maybe you've meant wether `id` might be required somewhere else outside of the scope of push notifications? Then no, AFAIK.